src/java.base/share/classes/jdk/internal/module/ModuleInfoWriter.java

```diff
@@ -46,122 +48,124 @@ public final class ModuleInfoWriter {
-
-        // provides
-        for (ModuleDescriptor.Provides p : md.provides()) {
-            mv.visitProvide(p.service().replace('.', '/'),
-                            p.providers()
-                                .stream()
-                                .map(pn -> pn.replace('.', '/'))
-                                .toArray(String[]::new));
+        try {
+            return Classfile.buildModule(
+                ModuleAttribute.of(ModuleDesc.of(md.name()), mb -> {
+                    mb.moduleFlags(modifiersToFlags(md.modifiers()));
+
+                    String vs = md.rawVersion().orElse(null);
+                    if (vs != null) mb.moduleVersion(vs);
+
+                    // requires
+                    for (ModuleDescriptor.Requires r : md.requires()) {
+                        int flags = r.modifiers().stream()
+                                .map(REQUIRES_MODS_TO_FLAGS::get)
+                                .reduce(0, (x, y) -> (x | y));
+                        vs = r.rawCompiledVersion().orElse(null);
+                        mb.requires(ModuleDesc.of(r.name()), flags, vs);
+                    }
+
+                    // exports
+                    for (ModuleDescriptor.Exports e : md.exports()) {
+                        int flags = e.modifiers().stream()
+                                .map(EXPORTS_MODS_TO_FLAGS::get)
+                                .reduce(0, (x, y) -> (x | y));
+                        var targets = e.targets().stream().map(ModuleDesc::of)
+                                .toArray(ModuleDesc[]::new);
+                        mb.exports(PackageDesc.of(e.source()), flags, targets);
+                    }
+
+                    // opens
+                    for (ModuleDescriptor.Opens opens : md.opens()) {
+                        int flags = opens.modifiers().stream()
+                                .map(OPENS_MODS_TO_FLAGS::get)
+                                .reduce(0, (x, y) -> (x | y));
+                        var targets = opens.targets().stream().map(ModuleDesc::of)
+                                .toArray(ModuleDesc[]::new);
+                        mb.opens(PackageDesc.of(opens.source()), flags, targets);
+                    }
+
+                    // uses
+                    md.uses().stream().map(sn -> ClassDesc.of(sn)).forEach(mb::uses);
+
+                    // provides
+                    for (ModuleDescriptor.Provides p : md.provides()) {
+                        mb.provides(ClassDesc.of(p.service()),
+                                        p.providers()
+                                            .stream()
+                                            .map(pn -> ClassDesc.of(pn))
+                                            .toArray(ClassDesc[]::new));
+                    }
+                }),
+
+                // packages
+                md.packages().stream().map(PackageDesc::of).toList(),
```

Comment:
When writing `module_info.class` via ASM, it only visits the packages if there are packages that aren't exported or open.

Does the ClassFile API add `ModulePackages` attribute only if there are packages that aren't exported or open?

Reply:
The ModulePackages attribute is optional and an optimization to avoid scanning the module contents to get the full set of packages. Tooling that creates packaged modules (jar and jmod for now) will want the ModulePackages attribute emitted always. So maybe the Classfile.buildModule methods should be looked at again, at least I think the 2-arg and 3-arg methods should emit the ModulePackages unconditionally, the 1-arg buildModule maybe not.

Reply:
`buildModules` is expected to be called with additional packages but instead it's called with all packages including all exported and open packages.   

```
    /**
     * Build a module descriptor into a byte array.
     * @param moduleAttribute the {@code Module} attribute
     * @param packages additional module packages
     * @param handler a handler that receives a {@link ClassBuilder}
     * @return the classfile bytes
     */
    public static byte[] buildModule(ModuleAttribute moduleAttribute,
                                     List<PackageDesc> packages,
                                     Consumer<? super ClassBuilder> handler) {
```

I checked the implementation that seems to match `@param packages` that expects additional module packages that are not exported nor open.  If it intends to take additional packages, it will need to filter the exported and open packages at the callsite.

Or the `packages` parameter lists all packages that will be used to create `ModulePackages` attribute.  This seems to be easier to understand.

Reply:
Maybe the variants of Classfile.buildModule need to be looked at again. For the usage here, buildModule(ModuleAttribute, Consumer<? super ClassBuilder>) would be more useful as it would allow all of the additional attributes to be emitted in the handler rather than having buildModule making the decision on whether to emit the ModulePackages attribute.

Reply:
`Classfile::buildModule` is a helper method simplifying the major use case - building valid module with minimal user effort and knowledge. It emits `ModulePackages` if there are any packages specified and not exported nor opened. That is the configuration which JVM expects and majority of users are not aware of.
Other module-specific attributes can be added in the underlying handler, using `ClassBuilder::with`.

A different building approach can be used in cases (which I would like to know more about) where even `ModulePackages` attribute needs to be "customised" or provided unconditionally. It is always possible to build module as a class using `Classfile::build` from scratch and provide all the attributes manually.

Reply:
Or another approach for the specific use cases might be to use `Classfile::buildModule` with empty list of packages and then provide custom `ModulePackagesAttribute` to the handler directly.

Reply:
I see, I'll move the "packages magic" back from Classfile API to the `ModuleInfoWriter` and correct the API accordingly.

Reply:
> I see, I'll move the "packages magic" back from Classfile API to the `ModuleInfoWriter` and correct the API accordingly.

I would need to check the history in the jake repo but I think it dates from a prototype version of that attribute that only listed the concealed packages. I think it should be okay to just drop this from ModuleInfoWriter now. As I mentioned in another comment, the only user now is tests so you should able to just remove this code so that it doesn't generate the ModulePackages attributes.

Reply:
> I think it would be better to change is so that calling it with ModuleAttribute will emit a Module attribute. The overloads that take a Module plus a set of packages can emit a Module + ModulePackage unconditionally.

I'm fine with this.

`buildModule(ModuleAttribute moduleAttribute, List<PackageDesc> packages, Consumer handler)` would emit `ModulePackages` unconditionally.   It means that it would need another `buildModule` variant that takes ModuleAttribute and handler only if no `ModulePackages` is generated.

Reply:
`buildModule` argument `List<PackageDesc> packages` was explicitly added only for the purpose of this conditional packages calculation. If the `ModulePackages` attribute is optional and its content is not calculated, there is no more reason for this argument. 
I would propose to reduce the method to `buildModule(ModuleAttribute moduleAttribute, Consumer handler)`. `ModulePackages` should be then built using regular approach inside the handler: `classBuilder.with(ModulePackagesAttribute.of(...));`.

Reply:
Taking out the variant with `<List<PackageDesc> packages` is a simplification.  Adding `ModulePackagesAttribute` is straight-forward.